rustls/src/msgs/handshake.rs

```diff
@@ -962,127 +962,157 @@ impl<'a> Codec<'a> for ClientSessionTicket {
-            Self::EarlyData => ExtensionType::EarlyData,
-            Self::EncryptedClientHello(_) => ExtensionType::EncryptedClientHello,
-            Self::Unknown(r) => r.typ,
+extension_struct! {
+    pub struct ServerExtensions<'a> {
+        /// Supported EC point formats (RFC4492)
+        ExtensionType::ECPointFormats =>
+            pub(crate) ec_point_formats: Option<Vec<ECPointFormat>>,
+
+        /// Server name indication acknowledgement (RFC6066)
+        ExtensionType::ServerName =>
+            pub(crate) server_name_ack: Option<()>,
+
+        /// Session ticket acknowledgement (RFC5077)
+        ExtensionType::SessionTicket =>
+            pub(crate) session_ticket_ack: Option<()>,
+
+        ExtensionType::RenegotiationInfo =>
+            pub(crate) renegotiation_info: Option<PayloadU8>,
+
+        /// Selected ALPN protocol (RFC7301)
+        ExtensionType::ALProtocolNegotiation =>
+            pub(crate) selected_protocol: Option<Vec<ProtocolName>>, // TODO: OneProtocolName
+
+        /// Key exchange server share (RFC8446)
+        ExtensionType::KeyShare =>
+            pub(crate) key_share: Option<KeyShareEntry>,
+
+        /// Selected preshared key index (RFC8446)
+        ExtensionType::PreSharedKey =>
+            pub(crate) preshared_key: Option<u16>,
+
+        /// Required client certificate type (RFC7250)
+        ExtensionType::ClientCertificateType =>
+            pub client_certificate_type: Option<CertificateType>,
+
+        /// Selected server certificate type (RFC7250)
+        ExtensionType::ServerCertificateType =>
+            pub server_certificate_type: Option<CertificateType>,
+
+        /// Extended master secret is in use (RFC7627)
+        ExtensionType::ExtendedMasterSecret =>
+            pub(crate) extended_master_secret_ack: Option<()>,
+
+        /// Certificate status acknowledgement (RFC6066)
+        ExtensionType::StatusRequest =>
+            pub(crate) certificate_status_request_ack: Option<()>,
+
+        /// Selected TLS version (RFC8446)
+        ExtensionType::SupportedVersions =>
+            pub(crate) selected_version: Option<ProtocolVersion>,
+
+        /// QUIC transport parameters (RFC9001)
+        ExtensionType::TransportParameters =>
+            pub(crate) transport_parameters: Option<Payload<'a>>,
+
+        /// QUIC transport parameters (RFC9001 prior to draft 33)
+        ExtensionType::TransportParametersDraft =>
+            pub(crate) transport_parameters_draft: Option<Payload<'a>>,
+
+        /// Early data is accepted (RFC8446)
+        ExtensionType::EarlyData =>
+            pub(crate) early_data_ack: Option<()>,
+
+        /// Encrypted inner client hello response (draft-ietf-tls-esni)
+        ExtensionType::EncryptedClientHello =>
+            pub(crate) encrypted_client_hello_ack: Option<ServerEncryptedClientHello>,
+    } + {
+        pub(crate) unknown_extensions: BTreeSet<u16>,
+    }
+}
+
+impl ServerExtensions<'_> {
```

Comment:
Couldn't the macro generate these `into_owned()` impls?

Reply:
Maybe theoretically? I don't seem to be smart enough to write such a macro though (the issue is replacing a `ty` matcher with something else which can match (eg) `Vec<T>` like normal, but also `T<'a>` and discern the two, such that the lifetime case can output `.map(|x| x.into_owned())`.)